ts/src/crypto/pedersen/pedersen.ts

```diff
@@ -0,0 +1,112 @@
+import { BarretenbergWasm } from "../../wasm/index.js";
+import {
+  deserializeArrayFromVector,
+  deserializeField,
+  serializeBufferArrayToVector,
+} from "../../wasm/serialize.js";
+import { Buffer } from "buffer";
+
+/**
+ * Combines two 32-byte hashes.
+ * @param wasm - The barretenberg module.
+ * @param lhs - The first hash.
+ * @param rhs - The second hash.
+ * @returns The new 32-byte hash.
+ */
+export function pedersenCompress(
+  wasm: BarretenbergWasm,
+  lhs: Uint8Array,
+  rhs: Uint8Array
+): Buffer {
+  // If not done already, precompute constants.
+  wasm.call("pedersen__init");
+  // TODO check if lhs and rhs are <= 32 bytes?
+  wasm.writeMemory(lhs, 0);
+  wasm.writeMemory(rhs, 32);
+  wasm.call("pedersen__compress_fields", 0, 32, 64);
+  return Buffer.from(wasm.getMemorySlice(64, 96));
+}
+
+/**
+ * Combine an array of hashes.
+ * @param wasm - The barretenberg module.
+ * @param lhs - The first hash.
+ * @param rhs - The second hash.
+ * @returns The new 32-byte hash.
+ */
+export function pedersenCompressInputs(
+  wasm: BarretenbergWasm,
+  inputs: Buffer[]
+): Buffer {
+  // If not done already, precompute constants.
+  wasm.call("pedersen__init");
+  const inputVectors = serializeBufferArrayToVector(inputs);
+  wasm.writeMemory(inputVectors, 0);
+  wasm.call("pedersen__compress", 0, 0);
+  return Buffer.from(wasm.getMemorySlice(0, 32));
+}
+
+/**
+ * Combine an array of hashes.
+ * @param wasm - The barretenberg module.
+ * @param lhs - The first hash.
+ * @param rhs - The second hash.
+ * @returns The new 32-byte hash.
+ */
+export function pedersenCompressWithHashIndex(
+  wasm: BarretenbergWasm,
+  inputs: Buffer[],
+  hashIndex: number
+): Buffer {
+  // If not done already, precompute constants.
+  wasm.call("pedersen__init");
+  const inputVectors = serializeBufferArrayToVector(inputs);
+  wasm.writeMemory(inputVectors, 0);
```

Comment:
Hm. Should we be writing to 0 instead of using bbmalloc?

Reply:
For data we can't control how big it will be should be written to allocated memory?

Reply:
My thought too. I did a straight refactor for now, but some stuff has overflow risks